src/main/kotlin/com/terraformation/backend/accelerator/db/ModuleEventStore.kt

```diff
@@ -0,0 +1,163 @@
+import java.net.URI
+import java.time.Duration
+import java.time.Instant
+import java.time.InstantSource
+import org.jooq.DSLContext
+import org.jooq.impl.DSL
+import org.springframework.context.ApplicationEventPublisher
+
+@Named
+class ModuleEventStore(
+    private val clock: InstantSource,
+    private val dslContext: DSLContext,
+    private val eventPublisher: ApplicationEventPublisher,
+    private val eventsDao: EventsDao,
+) {
+  private val eventProjectsHelper: ModuleEventProjectsHelper by lazy { ModuleEventProjectsHelper() }
+
+  fun fetchOneById(eventId: EventId, projectId: ProjectId? = null): EventModel {
+    requirePermissions { readModuleEvent(eventId) }
+
+    val projectsField =
+        if (projectId == null) {
+          eventProjectsHelper.eventProjectsMultiset()
+        } else {
+          requirePermissions { readModuleEvent(eventId) }
+          null
+        }
+
+    val projectEventCondition =
+        projectId?.let {
+          EVENTS.ID.`in`(
+              DSL.select(EVENT_PROJECTS.EVENT_ID)
+                  .from(EVENT_PROJECTS)
+                  .where(EVENT_PROJECTS.PROJECT_ID.eq(it)))
+        }
+
+    return with(EVENTS) {
+      dslContext
+          .select(asterisk(), projectsField)
+          .from(this)
+          .where(ID.eq(eventId))
+          .and(projectEventCondition)
+          .fetchOne { EventModel.of(it, projectsField) } ?: throw EventNotFoundException(eventId)
+    }
+  }
+
+  fun create(
+      moduleId: ModuleId,
+      eventType: EventType,
+      startTime: Instant? = null,
+      endTime: Instant? = startTime?.plus(Duration.ofHours(1)),
+      meetingUrl: URI? = null,
+      recordingUrl: URI? = null,
+      slidesUrl: URI? = null,
+  ): EventModel {
+    requirePermissions { manageModuleEvents() }
+
+    val now = clock.instant()
+    val user = currentUser()
+
+    val eventsRow =
+        EventsRow(
+            moduleId = moduleId,
+            eventTypeId = eventType,
+            startTime = startTime,
+            endTime = endTime,
+            meetingUrl = meetingUrl,
+            recordingUrl = recordingUrl,
+            revision = 1,
+            slidesUrl = slidesUrl,
+            createdBy = user.userId,
+            createdTime = now,
+            modifiedBy = user.userId,
+            modifiedTime = now,
+        )
+
+    eventsDao.insert(eventsRow)
+
+    val model = eventsRow.toModel()
+    startTime?.let {
+      eventPublisher.publishEvent(ModuleEventScheduledEvent(model.id, model.revision))
+    }
+
+    return model
+  }
+
+  fun updateEvent(eventId: EventId, updateFunc: (EventModel) -> EventModel) {
+    requirePermissions { manageModuleEvents() }
+
+    val existing = fetchOneById(eventId)
+    val updated = updateFunc(existing)
+    val userId = currentUser().userId
+    val now = clock.instant()
+
+    val revision =
+        if (updated.startTime == existing.startTime) {
+          existing.revision
+        } else {
+          existing.revision + 1
+        }
```

Comment:
I don't think it's likely enough to be worth worrying about, but just noting that this will produce unpredictable results if two different updates of the same event happen at the same time. (They'll both see the same existing revision number.)

Reply:
Should we look at a locking mechanism, or maybe is there a PostgreSQL atomic increment query that handles that on the DB side?

Reply:
There are a few ways you could do this if you wanted.

1. Locking, like you said. Start the transaction right away and do `SELECT FOR UPDATE` to read the events row.
2. Add a condition like `.and(REVISION.eq(existing.revision))` to the update. Then if the revision doesn't match, the update won't do anything. In that case, you will no longer be able to assume that the update touching 0 rows means the event doesn't exist.
3. Let the database do the incrementing for you: `.set(REVISION, REVISION.plus(1))` and then instead of `.execute()`, you'll want `.returning(REVISION).fetchOne()` to get back the updated revision.

That said, I wouldn't spend a lot of effort on this unless you want to do it as an exercise or something. Only admins can edit events and I think it's sufficiently unlikely there'd be two different updates to the same event at the exact same time that handling it is pretty low priority.